packages/server/src/fhir/operations/resourcegraph.ts

```diff
@@ -0,0 +1,333 @@
+    throw new OperationOutcomeError(badRequest('Invalid link path. Must return a path to a Reference type'));
+  }
+
+  // For each of the elements we found on the current resource, follow their various targets
+
+  const referenceElements = elements.filter((elem) => elem.type === PropertyType.Reference);
+  if (referenceElements.length > 0) {
+    results.push(...(await followReferenceElements(referenceElements, target, resourceCache, repo)));
+  }
+
+  const canonicalElements = elements.filter((elem) => elem.type === PropertyType.canonical);
+  if (canonicalElements.length > 0) {
+    results.push(...(await followCanonicalElements(canonicalElements, target, resourceCache, repo)));
+  }
+
+  return results;
+}
+
+async function followReferenceElements(
+  elements: TypedValue[],
+  target: GraphDefinitionLinkTarget,
+  resourceCache: Record<string, Resource>,
+  repo: Repository
+): Promise<Resource[]> {
+  const targetReferences = elements
+    .filter((elem) => elem.value.reference?.split('/')[0] === target.type)
+    .map((elem) => elem.value as Reference);
+
+  const results = [] as Resource[];
+
+  for (const ref of targetReferences) {
+    if (ref.reference) {
+      if (ref.reference in resourceCache) {
+        results.push(resourceCache[ref.reference]);
+      } else {
+        const linkedResource = await repo.readReference(ref);
+
+        // Cache here to speed up subsequent loop iterations
+        addToCache(linkedResource, resourceCache);
+        results.push(linkedResource);
+      }
+    }
+  }
+  return results;
+}
+
+async function followCanonicalElements(
+  elements: TypedValue[],
+  target: GraphDefinitionLinkTarget,
+  resourceCache: Record<string, Resource>,
+  repo: Repository
+): Promise<Resource[]> {
+  if (!target?.type) {
+    return [];
+  }
+
+  // Filter out Resources where we've seen the canonical URL
+  const targetUrls = elements.map((elem) => elem.value as string);
+
+  const results = [] as Resource[];
+  for (const url of targetUrls) {
+    if (url in resourceCache) {
+      results.push(resourceCache[url]);
+    } else {
+      const bundle = (await repo.search({
+        resourceType: target.type,
+        filters: [{ code: 'url', operator: Operator.EQUALS, value: url }],
+      })) as Bundle;
+      const linkedResources = bundle?.entry?.map((entry) => entry.resource).filter((e) => !!e) as Resource[];
+      if (linkedResources?.length > 1) {
+        console.warn(`Warning: Found more than 1 resource with canonical URL ${url}`);
+      }
+
+      // Cache here to speed up subsequent loop iterations
+      linkedResources.forEach((res) => addToCache(res, resourceCache));
+      results.push(...linkedResources);
+    }
+  }
+
+  return results;
+}
+
+async function followSearchLink(
+  resource: Resource,
+  link: GraphDefinitionLink,
+  resourceCache: Record<string, Resource>,
+  repo: Repository
```

Comment:
We try to use the convention that `repo` should always be the first argument to any server-side function requiring a `repo`.